File: Week_预习/G20200343040338/two-sum.py
# 解法1：遍历
class Solution1:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i, j]

# 解法2：使用字典（两遍字典法）
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        map = {num: idx for idx, num in enumerate(nums)}

        for idx, num in enumerate(nums):
            complement = target - num

            if complemnt in map and map[complement] != idx:
                return [idx, map[complement]]


# 解法3：一遍字典法
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        map = {}

        for idx, num in enumerate(nums):
            complement = target - num

            if complement in map:
                return [idx, map[complement]]

            map[num] = idx

File: Week_01/G20200343040156/twoSum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {

        if(null ==nums||nums.length ==0) {return null;}

        Map<Integer,Integer>  keyValueMap =new HashMap<>();

        for(int i=0;i<nums.length;++i){
            int currentValue = target-nums[i];
            if (keyValueMap.containsKey(currentValue)) {
                return new int[]{keyValueMap.get(currentValue),i};
            }
            keyValueMap.put(nums[i],i);
        }
        return null;
    }
}